### PR TITLE
Reusable table component and start of account settings

### PR DESCRIPTION
**Pull request by hrjuggins**
Created reusable table component (mainly to test out typescript) but should hopefully be useful

Made a start on account settings

** add relative path to tsconfig

### CODE REVIEW

File: src/components/Table/Table.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    width?: number;
+}
+
+type TableType<T, K extends keyof T> = {
+    columns: Array<ColumnDefinitionType<T, K>>,
+    data: Array<T>
+}
+
+const Table = <T, K extends keyof T>({
+	columns,
```
Comment (PH86) [inline src/components/Table/Table.tsx:21]: Indentation formatting

File: src/components/Table/components/TableHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+const TableHeader = <T, K extends keyof T>({columns} : TableHeaderProps<T, K>): React.ReactElement => {
+    const headers = columns.map((column, i) => {
+        const style = {
+            width: column.width ?? 100,
+        }
+        return (
+            <th
+            key={`th-${i}`}
```
Comment (PH86) [inline src/components/Table/components/TableHeader.tsx:15]: indentation formatting

File: src/pages/Account/Account.css [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+.section-container {
+     margin: 0.5em 0 1.5em;
+     display: flex; 
+     align-items: flex-start;
+     justify-content: space-between;
+     flex-wrap: wrap;
+}
+
+
```
Comment (PH86) [inline src/pages/Account/Account.css:14]: extra line to remove

File: src/pages/Account/Account.css [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    font-size: 2em;
+}
+
+.stat-box p {
+    margin-left: 0.5em;
+    font-size: 1em;
+}
+
+
+
```
Comment (PH86) [inline src/pages/Account/Account.css:73]: extra lines to remove

File: src/pages/Account/Account.tsx [hunk 5]
```diff
@@ -10,7 +60,34 @@ export const Account: React.FC<{}> = (): React.ReactElement => {
             variants={pageTransitions}
             className="content-container"
         >
-            <h1>Account</h1>
+            <h1>Account details</h1>
+
```
Comment (PH86) [inline src/pages/Account/Account.tsx:64]: empty line to remove

File: src/pages/Account/Account.tsx [hunk 6]
```diff
@@ -10,7 +60,34 @@ export const Account: React.FC<{}> = (): React.ReactElement => {
...
+            <div className="section-container">
+                <div className="detail-box sub-box">
+                    <p>Current subscription package</p>
+                    <h2>£25</h2>
+                    <p className="package-name">Standard Package</p>
+                    <button>Change Plan</button>
+                </div>
+                <button className="latest-bill">View Latest Bill</button>
+            </div>
+
```
Comment (PH86) [inline src/pages/Account/Account.tsx:74]: empty line to remove

File: src/components/Table/Table.css [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  }
+  .table td {
+    border-bottom: 1px solid var(--clr-secondary-grey);
+  }
+
+  .table tr:hover {
+    cursor: pointer;
+    background-color: var(--clr-primary-blue);
+    color: var(--clr-text-secondary)
+  }
```
Comment (JackHewitt08) [inline src/components/Table/Table.css:29]: new line

File: src/pages/Account/Account.tsx [hunk 8]
```diff
@@ -10,7 +60,34 @@ export const Account: React.FC<{}> = (): React.ReactElement => {
...
+                <div className="detail-box stat-box">
+                    <h2>12</h2>
+                    <p>Applications</p>
+                </div>
+                <div className="detail-box stat-box">
+                    <h2>2</h2>
+                    <p>Responses</p>
+                </div>
+            </div>
+
```
Comment (PH86) [inline src/pages/Account/Account.tsx:85]: empty line to remove

File: src/pages/Account/Account.tsx [hunk 9]
```diff
@@ -10,7 +60,34 @@ export const Account: React.FC<{}> = (): React.ReactElement => {
...
+                    <h2>2</h2>
+                    <p>Responses</p>
+                </div>
+            </div>
+            
+            <div className="bills-container">
+                <h3>Billing Summary</h3>
+                <Table columns={columns} data={billingData} />
+            </div>
+
```
Comment (PH86) [inline src/pages/Account/Account.tsx:90]: empty line to remove

File: src/components/Table/Table.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { TableHeader, TableRows } from './components';
+import './Table.css'
+
+// <T, K extends keyof T>
+// T means any type is ok and K is used as a type for our key property
```
Comment (JackHewitt08) [inline src/components/Table/Table.tsx:6]: I like this

File: src/components/Table/components/TableHeader.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        );
+    })
+    return (
+        <thead>
+            <tr>{headers}</tr>
+        </thead>
+    )
+}
+
+export default TableHeader;
```
Comment (JackHewitt08) [inline src/components/Table/components/TableHeader.tsx:29]: new line

File: src/components/Table/components/TableRows.tsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { ColumnDefinitionType } from '../Table'
+
+type TableRowsProps<T, K extends keyof T> = {
+    data: Array<T>;
+    columns: Array<ColumnDefinitionType<T, K>>;
+}
+
+const TableRows = <T, K extends keyof T>({data, columns}: TableRowsProps<T, K>): React.ReactElement => {
+    const rows = data.map((row, i) => {
```
Comment (JackHewitt08) [inline src/components/Table/components/TableRows.tsx]: in this instance i would prefer 'i' and 'j' to be named more explicitly, such as rowIndex, and columnIndex etc, just for readability in the future

File: src/components/Table/components/TableRows.tsx [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    });
+    
+    return (
+        <tbody>
+            {rows}
+        </tbody>
+    )
+}
+
+export default TableRows;
```
Comment (JackHewitt08) [inline src/components/Table/components/TableRows.tsx:31]: new line

File: src/components/Table/components/index.tsx [hunk 14]
```diff
@@ -0,0 +1,2 @@
+export {default as TableHeader} from './TableHeader'
+export {default as TableRows} from './TableRows'
```
Comment (JackHewitt08) [inline src/components/Table/components/index.tsx:2]: new line

File: src/components/Table/components/TableRows.tsx [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+import { ColumnDefinitionType } from '../Table'
+
+type TableRowsProps<T, K extends keyof T> = {
+    data: Array<T>;
+    columns: Array<ColumnDefinitionType<T, K>>;
+}
+
+const TableRows = <T, K extends keyof T>({data, columns}: TableRowsProps<T, K>): React.ReactElement => {
+    const rows = data.map((row, i) => {
+        return (
```
Comment (JackHewitt08) [inline src/components/Table/components/TableRows.tsx:11]: new line before return

File: src/components/Table/components/TableHeader.tsx [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+type TableHeaderProps<T, K extends keyof T> = {
+    columns: Array<ColumnDefinitionType<T, K>>;
+}
+
+const TableHeader = <T, K extends keyof T>({columns} : TableHeaderProps<T, K>): React.ReactElement => {
+    const headers = columns.map((column, i) => {
+        const style = {
+            width: column.width ?? 100,
+        }
+        return (
```
Comment (JackHewitt08) [inline src/components/Table/components/TableHeader.tsx:13]: new line before return

File: src/components/Table/Table.tsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+type TableType<T, K extends keyof T> = {
+    columns: Array<ColumnDefinitionType<T, K>>,
+    data: Array<T>
+}
+
+const Table = <T, K extends keyof T>({
+	columns,
+    data
+}: TableType<T, K>): React.ReactElement => {
+
```
Comment (JackHewitt08) [inline src/components/Table/Table.tsx:24]: extra line here not required

File: src/components/Table/components/TableHeader.tsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+            <th
+            key={`th-${i}`}
+                style={style}
+            >
+                {column.header}
+            </th>
+        );
+    })
+    return (
```
Comment (JackHewitt08) [inline src/components/Table/components/TableHeader.tsx:22]: new line before return

File: src/pages/Account/Account.css [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    border: none;
+    border-radius: 5px;
+    background-color: #33add2;
+    color: #ffffff;
+}
+
+.latest-bill {
+    background-color: var(--clr-button-primary);
+    color: var(--clr-text-secondary);
+}
```
Comment (JackHewitt08) [inline src/pages/Account/Account.css:86]: new line at end of file

File: src/pages/Account/Account.tsx [hunk 20]
```diff
@@ -1,5 +1,55 @@
 import { motion } from "framer-motion";
-import { pageTransitions } from "../../utils/Animations";
+import { pageTransitions } from "utils/Animations";
+import './Account.css'
+import Table, { ColumnDefinitionType } from 'components/Table/Table'
+
+interface Bill {
```
Comment (JackHewitt08) [inline src/pages/Account/Account.tsx]: Hungarian notation for interfaces please

File: src/pages/Account/Account.tsx (hunk 5 above)
Comment (JackHewitt08) [inline src/pages/Account/Account.tsx:64]: remove extra line

File: src/pages/Account/Account.tsx (hunk 6 above)
Comment (JackHewitt08) [inline src/pages/Account/Account.tsx:74]: remove extra line

File: src/pages/Account/Account.tsx (hunk 8 above)
Comment (JackHewitt08) [inline src/pages/Account/Account.tsx:85]: remove extra line

File: tsconfig.json [hunk 21]
```diff
@@ -1,5 +1,11 @@
 {
   "compilerOptions": {
+    "baseUrl": "src",
+    // "paths": {
```
Comment (JackHewitt08) [inline tsconfig.json]: no commented code

Comment (JackHewitt08) [reply tsconfig.json]: (that arent comments)

**Review by JackHewitt08 — CHANGES_REQUESTED** [review]

**Review by PH86 — APPROVED** [review]

**Review by sambuckley1995 — APPROVED** [review]

**Review by JackHewitt08 — APPROVED** [review]

**Pull request merged**